services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/modification/InitiateCancelSubscriptionHandler.java

```diff
@@ -0,0 +1,277 @@
+                getTypeFactory().get(ModifySubscriptionResponse.class);
+        response.setSubscription(subscriptionWithItems);
+        response.setCart(cart);
+        return response;
+    }
+
+
+    protected boolean canCancelSubscription(
+            @lombok.NonNull SubscriptionWithItems subscriptionWithItems,
+            @Nullable ContextInfo contextInfo) {
+        Assert.notNull(subscriptionWithItems.getSubscription(),
+                "SubscriptionWithItems must not be null");
+        Subscription subscription = subscriptionWithItems.getSubscription();
+        Assert.notNull(subscription, "Subscription must not be null");
+
+        if (subscription.getCancellationPolicyRef() == null) {
+            return false;
+        }
+
+        CancellationPolicy cancellationPolicy = cancellationPolicyProvider
+                .readById(subscription.getCancellationPolicyRef(), contextInfo);
+        subscriptionWithItems.setCancellationPolicy(cancellationPolicy);
+        String paymentStrategy = subscription.getPaymentStrategy();
+        if (cancellationPolicy.getCancellationPolicyDetail(paymentStrategy) == null) {
+            return false;
+        }
+
+        return cancellationPolicy.getCancellationPolicyDetail(paymentStrategy)
+                .isAllowCancellation();
+    }
+
+
+    protected CreateCartRequest buildCreateCartRequest(
+            @lombok.NonNull ModifySubscriptionRequest request,
+            @Nullable ContextInfo contextInfo) {
+        CreateCartRequest createCartRequest = getTypeFactory().get(CreateCartRequest.class);
+
+        createCartRequest.setType(DefaultCartTypes.STANDARD.name());
+        createCartRequest.setName(generateCartName(request.getSubscription().getSubscription()));
+
+        AddItemRequest cancelItem = buildAddItemRequestForCancelItem(request);
+        createCartRequest.getAddItemRequests().add(cancelItem);
+
+        List<AddItemRequest> cancellationFeeProducts =
+                buildAddItemRequestsForCancellationFeeProducts(request);
+
+        createCartRequest.getAddItemRequests().addAll(cancellationFeeProducts);
+        createCartRequest.setPriceCartRequest(
+                buildPriceCartRequest(request.getSubscription().getSubscription(), contextInfo));
+
+        return createCartRequest;
+    }
+
+    protected AddItemRequest buildAddItemRequestForCancelItem(
+            @lombok.NonNull ModifySubscriptionRequest request) {
+        AddItemRequest cancelItem = getTypeFactory().get(AddItemRequest.class);
+        cancelItem.setItemType(DefaultCartItemTypes.CUSTOM_ITEM.name());
+        cancelItem.setDiscountable(
+                subscriptionCancellationProperties.isDiscountAllowedForCancellationCharge());
+        cancelItem
+                .setTaxable(subscriptionCancellationProperties.isTaxAllowedForCancellationCharge());
+        cancelItem.setQuantity(1);
+        cancelItem.getNonCatalogItemDetails().setItemName(CANCELLATION_REQUEST);
```

Comment:
Also looks like we have to set a price here, we'll just need to overwrite it later in Cart Ops
```suggestion
        cancelItem.getNonCatalogItemDetails().setItemName(CANCELLATION_REQUEST);
        cancelItem.getNonCatalogItemDetails().setPrice(
                MonetaryUtils.zero(request.getSubscription().getSubscription().getCurrency()));
```

Reply:
Scratch this. We probably need to do something else.

Reply:
Actually we do want this.